idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/ui/dialogs/DialogSyntax.kt

```diff
@@ -0,0 +1,106 @@
+    fileIcon: Icon? = KotlinFileType.INSTANCE.icon,
+    actionName: (directory: PsiDirectory, newName: String, templateName: String) -> String =
+      { _, _, _ -> createText },
+    createFileFromTemplate: (name: String, template: FileTemplate, dir: PsiDirectory) -> PsiFile? =
+      { name, template, dir -> NewKotlinFileAction.createFileFromTemplateWithStat(name, template, dir) },
+    startInWriteAction: Boolean = false
+  ): ExtensionPhase = // TODO: Investigate system when there is no `postProcess`
+    addAnAction(actionId, newFileAction(createText, actionDescription, buildDialog, postProcess, fileIcon, actionName, createFileFromTemplate, startInWriteAction))
+
+  fun DialogSyntax.newFileAction(
+    createText: String,
+    actionDescription: String,
+    buildDialog: CreateFileFromTemplateDialog.Builder.(project: Project, directory: PsiDirectory) -> Unit = Noop.effect3,
+    postProcess: (createdElement: PsiFile, templateName: String, customProperties: MutableMap<String, String>) -> Unit = Noop.effect3,
+    fileIcon: Icon? = KotlinFileType.INSTANCE.icon,
+    actionName: (directory: PsiDirectory, newName: String, templateName: String) -> String =
+      { _, _, _ -> createText },
+    createFileFromTemplate: (name: String, template: FileTemplate, dir: PsiDirectory) -> PsiFile? =
+      { name, template, dir -> NewKotlinFileAction.createFileFromTemplate(name, template, dir) },
+    startInWriteAction: Boolean = false
+  ): CreateFileFromTemplateAction =
+    object : CreateFileFromTemplateAction(createText, actionDescription, fileIcon), DumbAware {
+      override fun getActionName(directory: PsiDirectory?, newName: String, templateName: String?): String? =
+        directory?.let { d -> templateName?.let { t -> actionName(d, newName, t) } }
+
+      override fun buildDialog(project: Project?, directory: PsiDirectory?, builder: CreateFileFromTemplateDialog.Builder?): Unit =
+        project?.let { p -> directory?.let { d -> builder?.let { b -> buildDialog(b, p, d) } } } ?: Unit
+
+      override fun startInWriteAction(): Boolean = startInWriteAction
+
+      override fun isAvailable(dataContext: DataContext?): Boolean =
+        dataContext?.let { ctx ->
+          if (super.isAvailable(ctx))
+            LangDataKeys.IDE_VIEW.getData(ctx)?.let { ideView: IdeView ->
+              PlatformDataKeys.PROJECT.getData(ctx)?.let { project ->
+                ideView.directories.any {
+                  ProjectRootManager.getInstance(project).fileIndex.isInSourceContent(it.virtualFile)
+                }
+              }
+            }
+          else
+            false
+        } ?: false
+
+      override fun postProcess(createdElement: PsiFile?, templateName: String?, customProperties: MutableMap<String, String>?): Unit =
+        createdElement?.let { el -> templateName?.let { name -> customProperties?.let { prop -> postProcess(el, name, prop) } } }
+          ?: Unit
```

Comment:
Is it necessary to return a `Unit`? I know we like to keep it pure and avoid Unit if possible but I was wondering about the motivation to do so over something this like this since we're just looking to make a call here:

```
override fun postProcess(createdElement: PsiFile?, templateName: String?, customProperties: MutableMap<String, String>?) =
         createdElement?.let { el -> 
                templateName?.let { name -> 
                      customProperties?.let { prop -> 
                            postProcess(el, name, prop) 
                      } 
               } 
         }
```

Reply:
Your function has a different return Type than the one this features needs.